rust_src/src/lists.rs

```diff
@@ -142,109 +268,150 @@ impl Iterator for TailsIter {
     fn next(&mut self) -> Option<Self::Item> {
         match self.tail.as_cons() {
             None => {
-                if self.errsym.is_some() && self.tail.is_not_nil() {
-                    wrong_type!(self.errsym.unwrap(), self.list)
+                if self.tail.is_not_nil() {
+                    if let Some(errsym) = self.errsym {
+                        wrong_type!(errsym, self.list);
+                    }
                 }
                 None
             }
-            Some(tail_cons) => {
-                self.tail = tail_cons.cdr();
-                self.q = self.q.wrapping_sub(1);
-                if self.q != 0 {
-                    if self.tail == self.tortoise {
-                        return self.circular();
-                    }
-                } else {
-                    self.n = self.n.wrapping_sub(1);
-                    if self.n > 0 {
-                        if self.tail == self.tortoise {
-                            return self.circular();
-                        }
-                    } else {
-                        self.max <<= 1;
-                        self.q = self.max as u16;
-                        self.n = self.max >> 16;
-                        self.tortoise = self.tail;
-                    }
+            Some(cons) => {
+                self.tail = cons.cdr();
+                match self.circular_checks {
+                    // when off we do not checks at all. When 'safe' the checks are performed
+                    // and the iteration exits but no errors are raised.
+                    LispConsCircularChecks::off => Some(cons),
+                    _ => self.check_circular(cons),
                 }
-                Some(tail_cons)
             }
         }
     }
 }
 
-pub struct TailsNoCircularCheckIter {
```

Comment:
This exists because sometimes Emacs will happily return a value that is not a cons cell. `nthcdr` is an example of this. The name will change, the code may change. I suspect I need it for the plist support as well.

Once I have the semantics nailed down I will rename this and add useful comments.

Reply:
TailsIter returns a LispCons. Which is an issue when there is not an available cons. Stopping the iteration there usually means an off by one style bug.